File: backend/src/services/currencyService.ts
import axios from 'axios';
import NodeCache from 'node-cache';
import logger from '../utils/logger';

// Set up cache with TTL of 24 hours (86400 seconds)
const currencyCache = new NodeCache({ stdTTL: 86400 });

// The currencies we currently support
export enum SupportedCurrency {
  USD = 'USD',
  DKK = 'DKK',
}

// The base currency used across the system (all conversions will be to/from this)
export const BASE_CURRENCY = SupportedCurrency.USD;

// Interface for currency rates
interface CurrencyRates {
  [currency: string]: number;
}

/**
 * Fetches latest currency conversion rates from an external API
 * This could be replaced with a paid API service for production
 */
async function fetchLatestRates(): Promise<CurrencyRates> {
  try {
    // For this example, we're using the free exchangerate-api.com service
    // In production, you might want to use a more reliable paid service
    const response = await axios.get('https://open.er-api.com/v6/latest/USD');

    if (response.data && response.data.rates) {
      logger.info('Successfully fetched currency rates');
      return response.data.rates;
    }
    throw new Error('Invalid response format from currency API');
  } catch (error) {
    logger.error('Failed to fetch currency rates:', error);
    // If API fails, return a fallback with hardcoded rates
    // This ensures the application can continue to function
    return {
      [SupportedCurrency.USD]: 1,
      [SupportedCurrency.DKK]: 6.8, // Fallback approximate value
    };
  }
}

/**
 * Get currency rates, from cache if possible, otherwise fetch fresh data
 */
async function getCurrencyRates(): Promise<CurrencyRates> {
  const cacheKey = 'currencyRates';
  const cachedRates = currencyCache.get<CurrencyRates>(cacheKey);

  if (cachedRates) {
    return cachedRates;
  }

  const freshRates = await fetchLatestRates();
  currencyCache.set(cacheKey, freshRates);
  return freshRates;
}

/**
 * Convert amount from one currency to another
 */
export async function convertCurrency(
  amount: number,
  fromCurrency: SupportedCurrency,
  toCurrency: SupportedCurrency
): Promise<number> {
  // If currencies are the same, no conversion needed
  if (fromCurrency === toCurrency) {
    return amount;
  }

  const rates = await getCurrencyRates();

  // Convert from source currency to USD (if not already USD)
  let amountInUSD = amount;
  if (fromCurrency !== SupportedCurrency.USD) {
    amountInUSD = amount / rates[fromCurrency];
  }

  // Convert from USD to target currency (if not USD)
  if (toCurrency === SupportedCurrency.USD) {
    return parseFloat(amountInUSD.toFixed(2));
  }

  const convertedAmount = amountInUSD * rates[toCurrency];
  return parseFloat(convertedAmount.toFixed(2));
}

/**
 * Force refresh the currency rates cache
 */
export async function refreshCurrencyRates(): Promise<CurrencyRates> {
  const cacheKey = 'currencyRates';
  const freshRates = await fetchLatestRates();
  currencyCache.set(cacheKey, freshRates);
  return freshRates;
}

/**
 * Get the list of supported currencies
 */
export function getSupportedCurrencies(): string[] {
  return Object.values(SupportedCurrency);
}

export default {
  convertCurrency,
  refreshCurrencyRates,
  getSupportedCurrencies,
};

File: backend/app/services/currency_service.py
# backend/app/services/currency_service.py
import requests
from typing import Dict, Optional
import logging
from cachetools import cached, TTLCache
from datetime import datetime

# Set up logging
logger = logging.getLogger(__name__)

# Cache for exchange rates, with 24-hour TTL
exchange_rate_cache = TTLCache(maxsize=100, ttl=86400)

class CurrencyService:
    """
    Centralized service for handling all currency conversions in the application
    All database values are stored in USD and converted only when needed
    """

    BASE_CURRENCY = "USD"

    @staticmethod
    @cached(cache=exchange_rate_cache)
    def get_exchange_rates(base_currency: str = "USD") -> Dict[str, float]:
        """
        Fetch exchange rates from an API with caching
        Uses USD as base currency by default
        """
        try:
            url = f"https://api.exchangerate-api.com/v4/latest/{base_currency}"
            response = requests.get(url, timeout=5)
            data = response.json()

            # Log successful fetching of rates
            logger.info(f"Successfully fetched exchange rates with base currency {base_currency}")

            # Return the rates or fallback
            return data.get('rates', {"USD": 1.0, "DKK": 6.8991310126})
        except Exception as e:
            logger.error(f"Error fetching exchange rates: {e}")
            # Fallback rates
            return {"USD": 1.0, "DKK": 6.8991310126}

    @staticmethod
    def convert_currency(
            amount: float,
            from_currency: str = "USD",
            to_currency: str = "USD",
            rates: Optional[Dict[str, float]] = None
    ) -> float:
        """
        Convert amount between currencies
        If rates are provided, uses them instead of fetching new ones
        """
        if amount == 0 or from_currency == to_currency:
            return amount

        # Get rates if not provided
        if not rates:
            rates = CurrencyService.get_exchange_rates("USD")

        # If base currency isn't USD, normalize to USD first
        amount_in_usd = amount
        if from_currency != "USD":
            if from_currency not in rates:
                logger.warning(f"Currency {from_currency} not found in rates, using 1.0")
                return amount
            amount_in_usd = amount / rates.get(from_currency, 1.0)

        # Then convert from USD to target currency
        if to_currency == "USD":
            return amount_in_usd

        if to_currency not in rates:
            logger.warning(f"Currency {to_currency} not found in rates, using USD value")
            return amount_in_usd

        return amount_in_usd * rates.get(to_currency, 1.0)

    @staticmethod
    def convert_monetary_values(
            data: Dict,
            from_currency: str,
            to_currency: str,
            fields: list
    ) -> Dict:
        """
        Convert all monetary values in a dictionary from one currency to another

        Args:
            data: Dictionary containing monetary values
            from_currency: Source currency
            to_currency: Target currency
            fields: List of field names to convert

        Returns:
            Dictionary with converted values
        """
        if from_currency == to_currency:
            return data

        # Get rates once to avoid multiple API calls
        rates = CurrencyService.get_exchange_rates("USD")

        # Clone the data to avoid modifying the original
        result = {**data}

        # Convert each field
        for field in fields:
            if field in result and isinstance(result[field], (int, float)):
                result[field] = CurrencyService.convert_currency(
                    result[field],
                    from_currency,
                    to_currency,
                    rates
                )

        return result

    @staticmethod
    def format_currency(
            amount: float,
            currency: str = "USD",
            locale: str = "en-US",
            decimal_places: int = 2
    ) -> str:
        """
        Format a currency amount according to the specified locale and currency
        """
        try:
            from babel.numbers import format_currency
            return format_currency(amount, currency, locale=locale)
        except ImportError:
            # Fallback if babel is not available
            symbol = {"USD": "$", "DKK": "kr"}.get(currency, currency)
            return f"{symbol} {amount:.{decimal_places}f}"